src/include/miopen/problem.hpp

```diff
@@ -0,0 +1,130 @@
+
+#pragma once
+
+#include <miopen/miopen.h>
+
+#include <miopen/convolution.hpp>
+#include <miopen/binary_serialization.hpp>
+#include <miopen/object.hpp>
+#include <miopen/operator_descriptor.hpp>
+#include <miopen/solver_id.hpp>
+#include <miopen/tensor.hpp>
+
+#include <cstring>
+#include <unordered_map>
+
+namespace miopen {
+
+struct Handle;
+struct Solution;
+struct SearchOptions;
+
+namespace conv {
+struct ProblemDescription;
+} // namespace conv
+
+struct Problem : miopenProblem
+{
+    Problem() = default;
+
+    const TensorDescriptor& GetTensorDescriptor(miopenTensorName_t name) const
+    {
+        return tensor_descriptors.at(name);
+    }
+
+    miopenProblemDirection_t GetDirection() const { return direction; }
+
+    bool RegisterTensorDescriptor(miopenTensorName_t name, TensorDescriptor descriptor)
+    {
+        return tensor_descriptors.emplace(std::make_pair(name, std::move(descriptor))).second;
+    }
+
+    void SetDirection(miopenProblemDirection_t value) { direction = value; }
+
+    void SetOperatorDescriptor(const OperatorDescriptor* descriptor)
+    {
+        operator_descriptor = std::shared_ptr<OperatorDescriptor>(descriptor->Clone());
+    }
+    const OperatorDescriptor& GetOperatorDescriptor() const { return *operator_descriptor; }
+
+    std::vector<Solution>
+    FindSolutions(Handle& handle, const SearchOptions& options, std::size_t max_solutions) const;
+
+    conv::ProblemDescription AsConvolution() const;
+
+    const TensorDescriptor& GetTensorDescriptorChecked(miopenTensorName_t name,
+                                                       const std::string& name_str) const;
+
+    template <class Stream, std::enable_if_t<IsBinarySerializationRelated<Stream>{}, bool> = true>
+    friend Stream& operator<<(Stream& stream, Problem& problem)
+    {
+        stream << problem.direction;
+        stream << problem.tensor_descriptors;
+
+        auto primitive = stream.IsDeserializing() ? solver::Primitive{}
+                                                  : problem.operator_descriptor->GetPrimitive();
+
+        stream << primitive;
+
+        switch(primitive)
+        {
+        case solver::Primitive::Convolution:
+            if(stream.IsDeserializing())
+                problem.operator_descriptor = std::make_shared<ConvolutionDescriptor>();
+            stream << *dynamic_cast<ConvolutionDescriptor*>(problem.operator_descriptor.get());
+            break;
+        case solver::Primitive::Activation:
+        case solver::Primitive::Batchnorm:
+        case solver::Primitive::Pooling:
+        case solver::Primitive::Invalid:
+        default: MIOPEN_THROW(miopenStatusNotImplemented);
+        }
+
+        return stream;
+    }
+
+private:
+    miopenProblemDirection_t direction = miopenProblemDirectionForward;
+    std::unordered_map<miopenTensorName_t, TensorDescriptor> tensor_descriptors;
+    std::shared_ptr<OperatorDescriptor> operator_descriptor;
```

Comment:
I think it would be better to store the descriptor as a variant. The number of descriptors used is close(whereas the serialized types is more open). 

Using a variant will simplify serialization as we dont need to pump serialization through some erased type which is more difficult.